Test Additions: Pre-Remove Role Usage outside of XDUser

ryanrath opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Before we start migrating Role* Tables to the equivalent Acl* Tables it's advantageous to have tests in place that ensure we are able to catch any changes in behavior in code that may be affected.

Some minor style cleanup / doc block additions were made to XDChartPool.php and XDReportManager.php. These changes were made as this Branch / PR was originally intended to contain the actual Role migrations and the first step to making those migrations was to clean up the file that contained the code to be migrated. I backed out the functional changes that were made once I switched to just providing additional tests but I left the style / doc block changes as they are not functional. If people would prefer, I can back out these changes and include them in a future PR. 

## Motivation and Context
Ensure that we're able to catch any unintended differences in behavior for code affected by Role migration. 

## Tests performed
End Points Tested:
  - POST internal_dashboard/controllers/mailer.php?operation=enum_target_addresses
  - POST /controllers/report_builder.php?operation=enum_available_charts
  - POST /controllers/report_builder.php?operation=enum_reports
  - POST /controllers/chart_pool.php?operation=add_to_queue
  - POST /controllers/chart_pool.php?operation=remove_from_queue
  - POST /controllers/report_builder.php?operation=enum_templates
  - POST /controllers/report_builder.php?operation=save_report
  - POST /controllers/report_builder.php?operation=remove_report_by_id
  - POST /controllers/report_builder.php?operation=get_new_report_name
  - POST /report_image_renderer.php
  - POST /internal_dashboard/controllers/controller.php?operation=enum_user_visits
  - POST /internal_dashboard/controllers/controller.php?operation=enum_user_visits_export
  - GET /rest/v1/users/current

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [X] Test Addition

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

jpwhite4 commented on open_xdmod/modules/xdmod/integration_tests/lib/TestHarness/TestParameterHelper.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $result = '';
+        for ($i = 0; $i < $max; $i++) {
+            $result .= rand(0, PHP_INT_MAX - 1) % 10;
```
Note that PHP_INT_MAX-1 is much bigger than getrandmax() on 64bit Linux php 5.4 so this function will not generate random numbers between 0 and 9. You should instead call:
```
rand(0,9);
```
if you want a random number between 0 and 9 (inclusive).

jpwhite4 commented on open_xdmod/modules/xdmod/integration_tests/lib/TestHarness/TestParameterHelper.php:
```diff
@@ -0,0 +1,73 @@
+    {
+        $result = array();
+        for ($i = 0; $i < $length; $i++) {
+            $result[] = self::$SOURCE[rand(0, PHP_INT_MAX - 1) % 62];
```
Same comment here as for line 69 - you should call `rand(0, 61)` to get a random number between 0 and 61 inclusive.

ryanrath commented:
@jpwhite4 I've made the changes you recommended in your code review & removed the unnecessary modifications for inclusion in other PR's. If you could re-review when possible that would be great. Thanks!

smgallo commented on open_xdmod/modules/xdmod/integration_tests/lib/Controllers/BaseUserAdminTest.php:
```diff
@@ -66,7 +67,11 @@ public function getTestFiles()
+            try {
+                self::removeUser($userId, $username);
+            } catch (Exception $e) {
+                echo "Exception removing user [$userId, $username] during teardown: [{$e->getCode()}] {$e->getMessage()}: \n{$e->getTraceAsString()}\n";
```
While this is fine in the test code, in the future I think that it is easier to read as:
```
echo sprintf(
    "Exception removing user [%d, %s] during teardown: [%d] %s:\n%s\n",
    $userId,
    $username,
    $e->getCode(),
    $e->getMessage(),
    $e->getTraceAsString()
);
```

smgallo commented on open_xdmod/modules/xdmod/integration_tests/lib/Controllers/BaseUserAdminTest.php:
```diff
@@ -105,15 +110,23 @@ protected static function removeUser($userId, $username)
+
+        if (strpos($actualMessage, 'user_does_not_exist') !== false) {
+            if ($actualSuccess !== false) {
+                throw new Exception("Remove User ['success'] Expected: false, Received: $actualSuccessMessage | $actualMessage");
```
You could clarify a bit so when someone reads the error its clearer. Something like:
sprintf(
    "Remove User: received '%s' but expected '%s'. Raw message: %d;
    $response[0]['success'],
    'false',
    $actualMessage
);

smgallo commented on open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UserAdminTest.php:
```diff
@@ -402,4 +402,206 @@ public function provideGetDwDescripters()
+            foreach ($expectedStats as $key => $expectedStat) {
+                $entryExists = $this->entryExists(
+                    $actualStats,
+                    function ($key, $value) use ($expectedStat, &$notFound, $expectedOutput) {
```
Doesn't look like `&$notFound`, `$expectedOutput` are used.

jpwhite4 commented on open_xdmod/modules/xdmod/integration_tests/lib/Controllers/ControllerTest.php:
```diff
@@ -515,6 +516,70 @@ function ($item) {
+        $expectedHttpCode = array_key_exists('http_code', $expected) ? $expected['http_code'] : 200;
+
+        foreach($testData as $key => $value) {
+            $testData[$key] = TestParameterHelper::processParam($value);
```
This preprocessing step should be done in the dataprovider function  not the test function since it is solely about generating the input and expected test data.

jpwhite4 commented on open_xdmod/modules/xdmod/integration_tests/lib/Controllers/UserAdminTest.php:
```diff
@@ -402,4 +402,206 @@ public function provideGetDwDescripters()
+
+            // Now we can get down to the business of comparing...
+            for ($i = 0; $i < $rows; $i++) {
+                if (isset($actual[$i]) && isset($expected[$i])) {
```
You need to check that the number of lines in the actual and expected are the same - if the actual has extra spurious rows then this test will still (incorrectly) pass.

ryanrath commented:
@smgallo @jpwhite4 Alrighty, I believe I've addressed the code review comments up to this point and resolved the issue with the tests borking on Shippable. Let me know if there's anything else you'd like me to address. Thanks!
